stats/thresholds.go

```diff
@@ -138,60 +252,54 @@ func NewThresholds(sources []string) (Thresholds, error) {
 }
 
 func newThresholdsWithConfig(configs []thresholdConfig) (Thresholds, error) {
-	rt := goja.New()
-	if _, err := rt.RunProgram(jsEnv); err != nil {
-		return Thresholds{}, fmt.Errorf("threshold builtin error: %w", err)
-	}
+	thresholds := make([]*Threshold, len(configs))
+	sinked := make(map[string]float64)
 
-	ts := make([]*Threshold, len(configs))
 	for i, config := range configs {
-		t, err := newThreshold(config.Threshold, rt, config.AbortOnFail, config.AbortGracePeriod)
+		t, err := newThreshold(config.Threshold, config.AbortOnFail, config.AbortGracePeriod)
 		if err != nil {
 			return Thresholds{}, fmt.Errorf("threshold %d error: %w", i, err)
 		}
-		ts[i] = t
+		thresholds[i] = t
 	}
 
-	return Thresholds{rt, ts, false}, nil
+	return Thresholds{thresholds, false, sinked}, nil
 }
 
-func (ts *Thresholds) updateVM(sink Sink, t time.Duration) error {
-	ts.Runtime.Set("__sink__", sink)
-	f := sink.Format(t)
-	for k, v := range f {
-		ts.Runtime.Set(k, v)
-	}
-	return nil
-}
-
-func (ts *Thresholds) runAll(t time.Duration) (bool, error) {
-	succ := true
-	for i, th := range ts.Thresholds {
-		b, err := th.run()
+func (ts *Thresholds) runAll(duration time.Duration) (bool, error) {
+	succeeded := true
+	for i, threshold := range ts.Thresholds {
+		b, err := threshold.run(ts.sinked)
 		if err != nil {
 			return false, fmt.Errorf("threshold %d run error: %w", i, err)
 		}
+
 		if !b {
-			succ = false
+			succeeded = false
 
-			if ts.Abort || !th.AbortOnFail {
+			if ts.Abort || !threshold.AbortOnFail {
 				continue
 			}
 
-			ts.Abort = !th.AbortGracePeriod.Valid ||
-				th.AbortGracePeriod.Duration < types.Duration(t)
+			ts.Abort = !threshold.AbortGracePeriod.Valid ||
+				threshold.AbortGracePeriod.Duration < types.Duration(duration)
 		}
 	}
-	return succ, nil
+
+	return succeeded, nil
 }
 
 // Run processes all the thresholds with the provided Sink at the provided time and returns if any
 // of them fails
-func (ts *Thresholds) Run(sink Sink, t time.Duration) (bool, error) {
-	if err := ts.updateVM(sink, t); err != nil {
-		return false, err
+func (ts *Thresholds) Run(sink Sink, duration time.Duration) (bool, error) {
+	// Update the sinks store
+	ts.sinked = make(map[string]float64)
+	f := sink.Format(duration)
```

Comment:
Looking at this with fresh eyes, I don't think custom percentiles are supported right now, are they? So here, `Format` in case of `Trend` will only calculate `P(90)` and `P(95)` by default. But what if user specifies something else? Like I just wrote in another comment:
```
http_req_duration: ['p(34)<123e-5']
````
It seems to me the old version was doing this in JS, with `jsEnvSrc` substitute...

Fixing this would likely require a cast to `TrendSink` as `P()` is not a part of `Sink` interface :slightly_frowning_face: I had to cast this in PRW extension too.

Reply:
That's a good question! I must admit, I focused on what ended up in the sinks, and not so much on how it ended in there. I'm going to run a bunch of tests, see what happens, and I'll let you know. Not only that, but I might ask you for help if things go south 🙏🏻

Reply:
yes it needs to support `p(<random float between 0 and 100>)`, and yes previously that was done through the "reexporting" of the Trend's [P method](https://github.com/grafana/k6/blob/14079775b9bad0b8b2f503a7fa1e787fbc5cba36/stats/sink.go#L113-L131)

Reply:
With my recent changes, the parser supports and verifies that supplied values are floats in the range `[0, 100]` in either integer or decimal notation. It doesn't support the *exponent* notation of floating numbers, though. Should it?

Reply:
It seems to me that this code still doesn't support percentiles?!?

I would also, looking at this code and the other places where `sink.Format` is used, I kind of want to remove this method. It's in most cases used as a way to get info about values for the given sink, but then we also cast the sink to its concrete type in order to get even more info from it. 

Given that there are 4 sink types, I feel like a switch in all cases (as the fields might not be the same) might be better :thinking: .

Especially as it turns out that the [CounterSink](https://github.com/grafana/k6/blob/bd221f75725dbf7b62590829eea9f49029d1508a/stats/sink.go#L61) does actually return a `rate` but [the summary code overwrites it](https://github.com/grafana/k6/blob/bd221f75725dbf7b62590829eea9f49029d1508a/js/summary.go#L57-L61) possibly with the same value :facepalm:  

And in the REST API where it is used "correctly" we have a [minor bug](https://github.com/grafana/k6/issues/1253) about Trends and also doesn't show the same stuff as in the summary :facepalm: .

Fixing all of this out of the scope for this PR, but I kind of feel it might be better to *not* depend on what `Format` returns in the thresholds

Reply:
I have opened a dedicated issue: #2320. Please don't hesitate to add information you find missing @MStoykov. Resolving this discussion; but let me know if I've missed something that still needs addressing 🙏🏻